### PR TITLE
test: add test for checkpoint timeout flushing

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem
https://github.com/neondatabase/neon/pull/6661 changed the layer flushing logic and led to OOMs in staging.
The issue turned out to be holding on to in-memory layers for too long. After OOMing we'd need to replay potentially
a lot of WAL.

## Summary of changes
Test that open layers get flushed after the `checkpoint_timeout` config and do not require WAL reingest upon restart.
The workload creates a number of timelines and writes some data to each, but not enough to trigger flushes via the `checkpoint_distance` config.

I ran this test against https://github.com/neondatabase/neon/pull/6661 and it was indeed failing.

Closes https://github.com/neondatabase/neon/issues/6945

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: test_runner/regress/test_pageserver_small_inmemory_layers.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    neon_env_builder: NeonEnvBuilder, immediate_shutdown: bool
+):
+    """
+    Test that open layers get flushed after the `checkpoint_timeout` config
+    and do not require WAL reingest upon restart.
+
+    The workload creates a number of timelines and writes some data to each,
+    but not enough to trigger flushes via the `checkpoint_distance` config.
+    """
+    env = neon_env_builder.init_start()
```
Comment (koivunej) [inline test_runner/regress/test_pageserver_small_inmemory_layers.py:89]:
```suggestion
    env = neon_env_builder.init_start()
```

While deflakying tests or trying to do that, I often drop the unused default tenant and timeline created by `init_start`.

Avoid creating the default tenant and timeline with 
```suggestion
    env = neon_env_builder.init_configs()
    env.start()
```

File: test_runner/regress/test_pageserver_small_inmemory_layers.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    tenant, timeline = env.neon_cli.create_tenant(conf=TENANT_CONF)
+    with env.endpoints.create_start("main", tenant_id=tenant) as ep:
+        conn = await ep.connect_async()
+        try:
+            await conn.execute("CREATE TABLE IF NOT EXISTS t(key serial primary key, value text)")
+            await conn.execute(
+                f"INSERT INTO t SELECT i, CONCAT('payload_', i) FROM generate_series(0,{entries}) as i"
+            )
+        finally:
+            await conn.close(timeout=10)
```
[the remark this answers is not in the document]
Comment (VladLazar) [reply test_runner/regress/test_pageserver_small_inmemory_layers.py:39]: I think they are ([blogpost](https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-3.html)), but we don't have one in the tree.

[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_pageserver_small_inmemory_layers.py:39]: Ok, I was expecting the `connect_async` returned conection to be one, but yeah, not important.

**Review by koivunej — APPROVED** [review]
Excellent that the reverted work fails this.

File: test_runner/regress/test_pageserver_small_inmemory_layers.py (hunk 1 above)
Comment (VladLazar) [reply test_runner/regress/test_pageserver_small_inmemory_layers.py:89]: Makes sense. The default tenant isn't used here. Will apply.

Comment (VladLazar) [reply test_runner/regress/test_pageserver_small_inmemory_layers.py]: https://github.com/neondatabase/neon/pull/6950/commits/e8b37666ca44692b2cd0fa0de7b4c6ae57545e08

**Review by problame — APPROVED** [review]

**Comment by problame** [comment]
Seems solid and non-flaky due to the large checkpoint timeout.

Maybe we can get away with a shorter timeout to avoid bloating CI times?
You can use the `pytest.mark.rerun` decorator before merging to run it a 100 times https://neondb.slack.com/archives/C033RQ5SPDH/p1708261053197239

**Comment by VladLazar** [comment]
> Seems solid and non-flaky due to the large checkpoint timeout.
> 
> Maybe we can get away with a shorter timeout to avoid bloating CI times? You can use the `pytest.mark.rerun` decorator before merging to run it a 100 times https://neondb.slack.com/archives/C033RQ5SPDH/p1708261053197239

Yeah, that's what I'm trying to do, but some other unit tests seems to be broken.

**Comment by VladLazar** [comment]
Seems stable now. I had to reduce the size of the workload.

**Pull request merged**